test/exporter/test_quantify_scheduler_exporter.py

```diff
@@ -76,26 +75,30 @@ def test_export(self):
             )
             assert mock_schedule.add.call_count == 7
 
-    def check_gate_not_supported(self, g: Gate) -> None:
-        register_manager = RegisterManager(QubitRegister(3))
-        ir = IR()
-        ir.add_gate(g)
+    @pytest.mark.parametrize(
+        "gate",
+        [
+            H(Qubit(0)),
+            SWAP(Qubit(0), Qubit(1)),
+            BlochSphereRotation(qubit=Qubit(0), axis=(1, 2, 3), angle=0.9876, phase=2.34),
+            CCZ(Qubit(0), Qubit(1), Qubit(2)),
+        ],
+        ids=["H", "SWAP", "BSR", "CCZ"],
+    )
+    def test_gates_not_supported(self, gate: Gate) -> None:
+        builder = CircuitBuilder(3)
+        builder.ir.add_gate(gate)
+        circuit = builder.to_circuit()
 
         with MockedQuantifyScheduler():
             with pytest.raises(Exception, match="Cannot export circuit: it contains unsupported gates"):
-                quantify_scheduler_exporter.export(Circuit(register_manager, ir))
-
-    def test_gates_not_supported(self):
-        self.check_gate_not_supported(H(Qubit(0)))
-        self.check_gate_not_supported(SWAP(Qubit(0), Qubit(1)))
-        self.check_gate_not_supported(BlochSphereRotation(qubit=Qubit(0), axis=(1, 2, 3), angle=0.9876, phase=2.34))
-        self.check_gate_not_supported(CCZ(Qubit(0), Qubit(1), Qubit(2)))
+                quantify_scheduler_exporter.export(circuit)
 
 
 @pytest.mark.skipif(
     importlib.util.find_spec("quantify_scheduler") is not None, reason="quantify_scheduler is installed"
 )
 def test_quantify_scheduler_not_installed() -> None:
-    empty_circuit = Circuit(RegisterManager(QubitRegister(1)), IR())
+    empty_circuit = CircuitBuilder(1).to_circuit()
```

Comment:
Why couldn't a circuit object be empty? `version 3` is a valid cQASM 3.0 program. The equivalent would be an empty circuit. Maybe the question here would be. Why is the number of qubits mandatory for a CircuitBuilder constructor?

Why would we implement `CircuitBuilder.decompose()` in the first place?

Reply:
This is not information for 'us', it is more for the usability. The output of a circuit builder not being a circuit is a bit strange from a user perspective.

Reply:
I don't fully understand what you mean, the output of the `CircuitBuilder` in the highlighted text is a `Circuit`. It contains a number of (qu)bits/wires, but has no operations defined on it.

Reply:
> Why couldn't a circuit object be empty? `version 3` is a valid cQASM 3.0 program. The equivalent would be an empty circuit. Maybe the question here would be. Why is the number of qubits mandatory for a CircuitBuilder constructor?
> 
> Why would we implement `CircuitBuilder.decompose()` in the first place?

Currently the number of qubits is static (unless you change the register attribute yourself, which currently is dangerous and error prone), so the best place to put the number of qubits is at the initialization. Keeping in mind that the number of qubits is static, I think it makes sense to require at least 1 qubit. While it is valid in cQASM to have no qubits, the `CircuitBuilder` is a higher level tool and I currently can't think of a useful situation were you would want a `CircuitBuilder` without any qubits, since you can't add any gates or measurements without having at least one qubit. Hence, a `CircuitBuilder` without any qubits cannot do anything except creating empty circuits.

Reply:
> I don't fully understand what you mean, the output of the `CircuitBuilder` in the highlighted text is a `Circuit`. It contains a number of (qu)bits/wires, but has no operations defined on it.

I think my comment surpasses that of this ticket. I don't like the `to_circuit()` method. I feel like a `CircuitBuilder` should be either implicitly mutated to a `Circuit` after each operation or the user should be properly warned if they accidently use the `CircuitBuilder` object as a `Circuit`, i.e they add all gates to a CircuitBuilder and then try to decompose without using the `to_circuit()` method before hand.

Reply:
The `CircuitBuilder` is a [design pattern](https://en.wikipedia.org/wiki/Builder_pattern). It separates the built entity, the `Circuit` from its builder, the `CircuitBuilder`. And the builder has to offer a method for retrieving the built entity. In this case it is called `to_circuit`, I think because the `to_<something>` is widely used in different parts of the code. We could call it `get_circuit`.

So, instead of letting a `Circuit` build itself (e.g., I'm an empty circuit to start with, then I add a Hadamard to myself, then I add a measure instruction...), you have a `CircuitBuilder`, and you tell that builder you want a Circuit with these many qubits, and then ask to add this and that, and at one point you ask for the result.

Anyway, the name `CircuitBuilder` should be clear enough to indicate that you are not a `Circuit` but a `CircuitBuilder`. When we have proper documentation, and an API explaining the methods in a `CircuitBuilder` and in a `Circuit`, a user will have an ever clearer reference.

Reply:
To me the `CircuitBuilder` can create the circuits and I agree with @rturrado that this is a widely used patter and a good practice to separate to building of a complex data structure from its use.

From a design perspective, we are currently using the circuit object only for transpilation purposes. I really like that the transpilation is disconnected from the creation of the circuit. Once we have a circuit object, we (should) only transpile it, i.e. the meaning of the circuit doesn't change. We (should) not add/remove gates that change the meaning at this point. This is different from `qiskit` and `pennylane`, but in my opinion this is a good thing. I think there should be a separation between the building stage and compilation stage. Otherwise you can do this:
```python
from qiskit import QuantumCircuit, transpile

circuit = QuantumCircuit(3)
circuit.ccx(0,1,2)
transpiled_circuit = transpile(circuit, basis_gates = ['rx', 'h', 'cx'])
assert type(circuit) == type(transpiled_circuit)
transpiled_circuit.ccx(0,1,2)
```
The `transpiled_circuit` is no longer transpiled, while this is not executable anymore without any warning. For the mapping and routing stage this could completely undo the work you have done before without raising any warning.

That being said, you should be able to build a `CircuitBuilder` from a `Circuit` so you can possibly extend a (transpiled) circuit if you want to.

Reply:
Thank you for the insight guys! I now understand the perspective of having the `Circuit` and `CircuitBuilder` from a transpilation perspective. I will resolve this comment chain but these insights should definitely be documented somewhere.  

> That being said, you should be able to build a CircuitBuilder from a Circuit so you can possibly extend a (transpiled) circuit if you want to.

Also I will add this as a ticket.

Reply:
I don't agree with this last statement from Stan.

> That being said, you should be able to build a CircuitBuilder from a Circuit so you can possibly extend a (transpiled) circuit if you want to.

I have already commented on the new ticket. I think that's an XY problem.

If the problem is that you may want to extend the circuit, one solution could be 1) not to lost track of the builder, and 2) that every call to `builder.to_circuit` would give you a new circuit.

You can not turn this into: I don't have access to the builder anymore, so I have to create it out of a circuit. That's a wrong solution for a wrong problem. You should think that creating mums out of children is wrong.